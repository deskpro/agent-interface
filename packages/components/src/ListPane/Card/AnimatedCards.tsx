import * as React from "react";
import { TransitionGroup, CSSTransition } from "react-transition-group";

export interface AnimatedItem {
  id: React.Key;
}

export type AnimatedCardsProps = {
  items: AnimatedItem[];
  children: (item: AnimatedItem, index: number) => React.ReactNode;
};

const AnimatedCards: React.FC<AnimatedCardsProps> = ({ items, children }) => (
  <TransitionGroup component={null}>
    {items.map((item, index) => (
      <CSSTransition
        classNames={{
          enter: "dp-Entering",
          enterActive: "dp-Entering-active",
          enterDone: "dp-Entering-done",
          exit: "dp-Leaving",
          exitActive: "dp-Leaving-active",
          exitDone: "dp-Leaving-done"
        }}
        timeout={500}
        appear={false}
        key={item.id}
      >
        {children(item, index)}
      </CSSTransition>
    ))}
  </TransitionGroup>
);

export default AnimatedCards;
