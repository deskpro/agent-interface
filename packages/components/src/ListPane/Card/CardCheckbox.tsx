import * as React from "react";

import Checkbox from "../../inputs/Checkbox/Checkbox";

export type CheckableCardProps = {
  cardId: React.Key;
  checkable?: boolean;
  checked?: boolean;
  onCheck: (cardId: React.Key, e: React.MouseEvent<HTMLInputElement>) => void;
};

const CardCheckbox: React.FC<CheckableCardProps> = ({
  cardId,
  checkable,
  checked,
  onCheck
}) => {
  const handleCheck = React.useCallback(
    (e: React.PointerEvent<HTMLInputElement>) => onCheck(cardId, e),
    [cardId, onCheck]
  );

  if (!checkable) {
    return null;
  }

  return (
    <Checkbox
      id={cardId.toString()}
      value={cardId}
      checked={checked}
      onChange={handleCheck}
    />
  );
};

export default CardCheckbox;
