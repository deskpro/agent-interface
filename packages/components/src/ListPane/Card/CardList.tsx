import * as React from "react";
import classNames from "classnames";

import "@deskpro/agent-interface-style/dist/components/dp-ListPane.css";

export type CardListProps = {
  className?: string;
  hoverable?: boolean;
  innerRef?: React.Ref<HTMLDivElement>;
  children: React.ReactNode;
};

const CardList: React.FC<CardListProps> = ({
  children,
  className,
  hoverable = false,
  innerRef
}) => (
  <div
    className={classNames("dp-CardList", className, {
      "is-hoverable": hoverable
    })}
    role="list"
    ref={innerRef}
  >
    {children}
  </div>
);

export default React.forwardRef<HTMLDivElement, CardListProps>((props, ref) => (
  <CardList {...props} innerRef={ref} />
));
