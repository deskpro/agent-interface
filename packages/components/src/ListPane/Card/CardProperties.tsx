import * as React from "react";

import Meta, { MetaProps } from "../../elements/Meta/Meta";

export type CardPropertiesProps = {
  properties: MetaProps[];
};

const CardProperties: React.FC<CardPropertiesProps> = ({ properties }) => (
  <span className="dp-LevelItem dp-LevelLeft">
    {properties.map(prop => (
      <Meta key={prop.name} {...prop} />
    ))}
  </span>
);

export default CardProperties;
