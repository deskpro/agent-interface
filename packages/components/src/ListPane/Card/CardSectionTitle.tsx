import * as React from "react";
import classNames from "classnames";

export type CardSectionTitleProps = {
  className?: string;
};

const CardSectionTitle: React.FC<CardSectionTitleProps> = ({
  className,
  children
}) => (
  <div
    className={classNames(
      "dp-Card-SectionTitle dp-SectionCardTitle",
      className
    )}
  >
    {children}
  </div>
);

export default CardSectionTitle;
