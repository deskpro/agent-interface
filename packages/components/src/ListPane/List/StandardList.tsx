import * as React from "react";

import List, { MassActionType } from "./List";
import { SimpleCardProps } from "../Card/SimpleCard";
import Pagination from "../../Pagination/Pagination";
import ActionsButton from "../../Button/ActionsButton";
import Menu from "../../elements/Menu/Menu";
import ListBar from "./ListBar";
import OrderingBar from "../OrderingBar/OrderingBar";
import Icon from "../../elements/Icon/Icon";
import { ListItem } from "../types";
import useItemSelection from "../useItemSelection";
import Arrow from "../../Button/Arrow";

export type StandardListProps = {
  className?: string;
  height?: number;
  items: ListItem[];
  numPages: number;
  massActions?: MassActionType[];
  onMassAction?: (actionName: React.Key, selectedItems: ListItem[]) => void;
  onLoadItems: (page: number) => void;
  showAddButton?: boolean;
  onAddClick?: (e: React.SyntheticEvent) => void;
  showPaginator?: boolean;
  renderItem: (
    item: ListItem,
    props: Partial<SimpleCardProps>
  ) => React.ReactNode;
};

const StandardList: React.FC<StandardListProps> = ({
  items,
  height,
  numPages,
  onLoadItems,
  massActions,
  onMassAction,
  showAddButton = false,
  onAddClick,
  showPaginator = true,
  ...props
}) => {
  const {
    selection,
    toggleItemSelection,
    toggleGroupSelection,
    getGroupCheckState,
    handleSelectAll,
    handleSelectNone,
    handleSelectInverse
  } = useItemSelection(items);

  // reset selection on new items.
  React.useEffect(handleSelectNone, [items]);

  const [currentPage, setCurrentPage] = React.useState(1);
  const handlePageChange = React.useCallback(
    nextPage => {
      setCurrentPage(nextPage);
      onLoadItems(nextPage);
    },
    [setCurrentPage, onLoadItems]
  );

  return (
    <>
      {showAddButton && onAddClick && (
        <List.AddButton onClick={onAddClick}>
          Click to add new item
        </List.AddButton>
      )}
      <ListBar
        totalItemsCount={items.length}
        selectedCount={selection.length}
        onSelectAll={handleSelectAll}
        onSelectNone={handleSelectNone}
        onSelectInverse={handleSelectInverse}
      >
        <OrderingBar.Item>
          <Icon name="grouped-by" size={19} />
          <span className="dp-SelectGroup">
            <span className="dp-ItemRow">
              <span className="dp-Title">Grouped by</span>
              <Arrow />
            </span>
            <span className="dp-ItemRow">
              <span className="dp-Subtitle">None</span>
            </span>
          </span>
        </OrderingBar.Item>
        <OrderingBar.Item>
          <Icon name="sort" size={19} />
          <span className="dp-SelectGroup">
            <span className="dp-ItemRow">
              <span className="dp-Title">Sort by</span>
              <Arrow />
            </span>
            <span className="dp-ItemRow">
              <span className="dp-Subtitle">Date</span>
            </span>
          </span>
        </OrderingBar.Item>
        <OrderingBar.Item>
          <Icon name="faceted-by" size={19} />
          <span className="dp-SelectGroup">
            <span className="dp-ItemRow">
              <span className="dp-Title">Faceted by</span>
              <Arrow />
            </span>
            <span className="dp-ItemRow">
              <span className="dp-Subtitle">3 Agents</span>
            </span>
          </span>
        </OrderingBar.Item>
        <OrderingBar.Item>
          <Icon name="fields" size={19} />
          <span className="dp-SelectGroup">
            <span className="dp-ItemRow">
              <span className="dp-Title">Fields</span>
              <Arrow />
            </span>
            <span className="dp-ItemRow">
              <span className="dp-Subtitle">8 fields</span>
            </span>
          </span>
        </OrderingBar.Item>
      </ListBar>
      <List
        {...props}
        key={currentPage}
        items={items}
        selected={selection}
        onSelectToggle={toggleItemSelection}
        onGroupSelectionToggle={toggleGroupSelection}
        getGroupCheckState={getGroupCheckState}
        scrollHeight={
          height
            ? height -
              (selection.length && massActions && massActions.length
                ? 95
                : 73) -
              (showAddButton ? 38 : 0)
            : undefined
        }
      />
      {showPaginator && (
        <Pagination
          currentPage={currentPage}
          numPages={numPages}
          onPageChange={handlePageChange}
          showGotoPage
        />
      )}
      {selection.length > 0 && massActions && massActions.length > 0 && (
        <ActionsButton
          renderMenu={menuProps => (
            <Menu {...menuProps}>
              {massActions.map(({ name, label }) => (
                <Menu.MenuItem
                  key={name}
                  name={name}
                  text={label}
                  onClick={() => {
                    if (onMassAction) {
                      onMassAction(
                        name,
                        items.filter(i => selection.includes(i.id))
                      );
                    }
                  }}
                />
              ))}
            </Menu>
          )}
        >
          Perform mass action
        </ActionsButton>
      )}
    </>
  );
};

export default StandardList;
