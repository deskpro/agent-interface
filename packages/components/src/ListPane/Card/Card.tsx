import * as React from "react";
import classNames from "classnames";

import CardTitle from "./CardTitle";
import CardSectionTitle from "./CardSectionTitle";
import CardList from "./CardList";
import TicketCard from "./TicketCard";
import TaskCard from "./TaskCard";
import { MenuProps } from "../../elements/Menu/Menu";
import Cog from "../../elements/Cog/Cog";
// import "@deskpro/agent-interface-style/dist/components/dp-ListPane.css";
import "@deskpro/agent-interface-style/dist/components/dp-Level.css";
import "@deskpro/agent-interface-style/dist/components/dp-card.css";

type CardSubComponents = {
  Title: typeof CardTitle;
  SectionTitle: typeof CardSectionTitle;
  List: typeof CardList;
  TicketCard: typeof TicketCard;
  TaskCard: typeof TaskCard;
};

export interface ICardModel {
  id: string | number;
}

export type CheckableCardProps = {
  checkable?: boolean;
  checked?: boolean;
  onCheck: (e: React.MouseEvent<HTMLInputElement>) => void;
};

export type CardProps<M> = {
  model: M;
  className?: string;
  isSelected: boolean;
  isActive?: boolean;
  isDragging?: boolean;
  title?: string | React.ReactNode;
  onClick?(event: React.MouseEvent<HTMLElement>): void;
  cogMenu?: React.ReactElement<MenuProps>;
  renderBody?(model: M): React.ReactNode;
};

const Card: React.FC<CardProps<ICardModel>> & CardSubComponents = ({
  model,
  className,
  children,
  isSelected = false,
  isActive = false,
  isDragging = false,
  title,
  cogMenu,
  renderBody
  // onClick
}) => (
  <div
    className={classNames("dp-Card dp-Level", className, {
      "is-selected": isSelected,
      "is-active": isActive,
      "is-dragging": isDragging
    })}
  >
    {!!title && <CardTitle title={title} />}
    {!!renderBody && renderBody(model)}
    {children}
    {!!cogMenu && <Cog menu={cogMenu} />}
  </div>
);

Card.Title = CardTitle;
Card.SectionTitle = CardSectionTitle;
Card.List = CardList;
Card.TicketCard = TicketCard;
Card.TaskCard = TaskCard;

export default Card;
