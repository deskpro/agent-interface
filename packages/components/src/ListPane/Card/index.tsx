import Card from "./Card";
import CardList from "./CardList";
import CardSectionTitle from "./CardSectionTitle";

Card.List = CardList;
Card.SectionTitle = CardSectionTitle;

export default Card;
