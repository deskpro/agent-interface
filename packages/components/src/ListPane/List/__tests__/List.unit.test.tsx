import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import StandardList from "../StandardList";
import Card from "../../Card";

describe("<List />", () => {
  it("should match its snapshot", () => {
    const handleCheckbox = jest.fn();
    const onLoadCallback = jest.fn();
    const { container } = render(
      <StandardList
        items={[
          { id: 1, title: "First item" },
          { id: 2, title: "Second Title" }
        ]}
        renderItem={(item, props) => (
          <Card.Simple
            {...props}
            title={item.title}
            cardId={item.id}
            onCheck={handleCheckbox}
          />
        )}
        onLoadItems={onLoadCallback}
        numPages={25}
      />
    );
    expect(container.querySelector(".dp-CardList")).toMatchSnapshot();
  });

  it("should call load callback on page clicks", () => {
    const handleCheckbox = jest.fn();
    const onLoadCallback = jest.fn();
    const { container } = render(
      <StandardList
        items={[
          { id: 1, title: "First item" },
          { id: 2, title: "Second Title" }
        ]}
        renderItem={(item, props) => (
          <Card.Simple
            {...props}
            title={item.title}
            cardId={item.id}
            onCheck={handleCheckbox}
          />
        )}
        onLoadItems={onLoadCallback}
        numPages={25}
      />
    );

    expect(container.querySelectorAll(".dp-PageNumber").length).toBe(8);

    fireEvent.click(container.querySelector("li.next a.dp-PaginationArrows"));
    expect(onLoadCallback).toBeCalledWith(2);
  });

  it("should change select_all checkbox state on item check", () => {
    const handleCheckbox = jest.fn();
    const onLoadCallback = jest.fn();
    const { container } = render(
      <StandardList
        items={[
          { id: "1ab", title: "First item" },
          { id: "2fe", title: "Second Title" }
        ]}
        renderItem={(item, props) => (
          <Card.Simple
            {...props}
            title={item.title}
            cardId={item.id}
            onCheck={handleCheckbox}
          />
        )}
        onLoadItems={onLoadCallback}
        numPages={25}
      />
    );

    expect(container.querySelector("input[value='2fe']")).toBeDefined();
    fireEvent.click(container.querySelector("input[value='2fe']"));
    expect(handleCheckbox.mock.calls[0][0]).toBe("2fe");
  });
});
