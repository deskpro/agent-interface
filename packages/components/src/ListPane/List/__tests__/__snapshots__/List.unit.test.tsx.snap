// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should change select_all checkbox state on item check 1`] = `
<div
  class="dp-OrderingBar"
>
  <div
    class="dp-OrderingBar-item is-expanded"
  >
    <span
      class="dp-Dropdown-container"
    >
      <span
        class="dp-Dropdown"
        role="button"
        tabindex="-1"
      >
        <span
          class="dp-CustomCheckbox"
        >
          <input
            class="dp-ControlInput"
            id="select_all"
            type="checkbox"
          />
          <label
            class=""
            for="select_all"
          />
        </span>
        <span
          class="dp-Icon dp-Arrow is-active"
          role="button"
          tabindex="-1"
        />
      </span>
      <ul
        class="dp-Menu is-visible"
        style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
      >
        <li
          class="dp-Menu-linkItem"
          role="menuitem"
        >
          <span
            style="display: inline-block; width: 100%;"
          >
            <span
              class="dp-Menu-link"
            >
              Select All
            </span>
          </span>
        </li>
        <li
          class="dp-Menu-linkItem"
          role="menuitem"
        >
          <span
            style="display: inline-block; width: 100%;"
          >
            <span
              class="dp-Menu-link"
            >
              Select none
            </span>
          </span>
        </li>
        <li
          class="dp-Menu-linkItem"
          role="menuitem"
        >
          <span
            style="display: inline-block; width: 100%;"
          >
            <span
              class="dp-Menu-link"
            >
              Inverse
            </span>
          </span>
        </li>
      </ul>
    </span>
    <span
      class="dp-MassSelect-itemsCount"
    >
      (2 items selected)
    </span>
  </div>
  <div
    class="dp-OrderingBar-item"
  >
    <span
      class="dp-SelectGroup"
    >
      <span
        class="dp-ItemRow"
      >
        <span
          class="dp-Title"
        >
          Grouped by
        </span>
        <span
          class="dp-Icon dp-Arrow"
          role="button"
          tabindex="-1"
        />
      </span>
      <span
        class="dp-ItemRow"
      >
        <span
          class="dp-Subtitle"
        >
          None
        </span>
      </span>
    </span>
  </div>
  <div
    class="dp-OrderingBar-item"
  >
    <span
      class="dp-SelectGroup"
    >
      <span
        class="dp-ItemRow"
      >
        <span
          class="dp-Title"
        >
          Sort by
        </span>
        <span
          class="dp-Icon dp-Arrow"
          role="button"
          tabindex="-1"
        />
      </span>
      <span
        class="dp-ItemRow"
      >
        <span
          class="dp-Subtitle"
        >
          Date
        </span>
      </span>
    </span>
  </div>
  <div
    class="dp-OrderingBar-item"
  >
    <span
      class="dp-SelectGroup"
    >
      <span
        class="dp-ItemRow"
      >
        <span
          class="dp-Title"
        >
          Faceted by
        </span>
        <span
          class="dp-Icon dp-Arrow"
          role="button"
          tabindex="-1"
        />
      </span>
      <span
        class="dp-ItemRow"
      >
        <span
          class="dp-Subtitle"
        >
          3 Agents
        </span>
      </span>
    </span>
  </div>
  <div
    class="dp-OrderingBar-item"
  >
    <span
      class="dp-SelectGroup"
    >
      <span
        class="dp-ItemRow"
      >
        <span
          class="dp-Title"
        >
          Fields
        </span>
        <span
          class="dp-Icon dp-Arrow"
          role="button"
          tabindex="-1"
        />
      </span>
      <span
        class="dp-ItemRow"
      >
        <span
          class="dp-Subtitle"
        >
          8 fields
        </span>
      </span>
    </span>
  </div>
</div>
`;

exports[`<List /> should match its snapshot 1`] = `
<div
  class="dp-CardList"
  role="list"
>
  <div
    class="dp-ScrollArea dp-List-items"
    style="position: relative; overflow: hidden; width: 100%;"
  >
    <div
      class="dp-ScrollArea-view"
      style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; overflow: scroll; margin-right: 0px; margin-bottom: 0px;"
    >
      <div
        class="dp-Card"
        role="link"
        tabindex="-1"
      >
        <div
          class="dp-LevelItem Level-justify-between"
        >
          <span
            class="dp-LevelLeft"
          >
            <span
              class="dp-CustomCheckbox dp-LevelTop"
            >
              <input
                class="dp-ControlInput"
                id="1"
                type="checkbox"
                value="1"
              />
              <label
                class=""
                for="1"
              />
            </span>
            <span
              class="dp-Card-title"
            >
              First item
            </span>
          </span>
        </div>
        <div
          class="dp-LevelItem Level-justify-between"
        >
          <span
            class="dp-LevelLeft"
          />
        </div>
      </div>
      <div
        class="dp-Card"
        role="link"
        tabindex="-1"
      >
        <div
          class="dp-LevelItem Level-justify-between"
        >
          <span
            class="dp-LevelLeft"
          >
            <span
              class="dp-CustomCheckbox dp-LevelTop"
            >
              <input
                class="dp-ControlInput"
                id="2"
                type="checkbox"
                value="2"
              />
              <label
                class=""
                for="2"
              />
            </span>
            <span
              class="dp-Card-title"
            >
              Second Title
            </span>
          </span>
        </div>
        <div
          class="dp-LevelItem Level-justify-between"
        >
          <span
            class="dp-LevelLeft"
          />
        </div>
      </div>
    </div>
    <div
      class="dp-ScrollArea-trackHorizontal"
      style="position: absolute; height: 6px; transition: opacity 200ms; opacity: 0; display: none;"
    >
      <div
        class="dp-ScrollArea-thumbHorizontal"
        style="position: relative; display: block; height: 100%;"
      />
    </div>
    <div
      class="dp-ScrollArea-trackVertical"
      style="position: absolute; width: 6px; transition: opacity 200ms; opacity: 0; display: none;"
    >
      <div
        class="dp-ScrollArea-thumbVertical"
        style="position: relative; display: block; width: 100%;"
      />
    </div>
  </div>
</div>
`;
